### PR TITLE
TinyBasic.py: Add language definition of TinyBasic

### PR DESCRIPTION
**Pull request by CLiu13**
This adds a new language definition
of TinyBasic, written in Python.

Closes https://github.com/coala/coala/issues/5824

**For short term contributors:** we understand that getting your commits well
defined like we require is a hard task and takes some learning. If you
look to help without wanting to contribute long term there's no need
for you to learn this. Just drop us a message and we'll take care of brushing
up your stuff for merge!

### Checklist

- [X] I read the [commit guidelines](http://coala.io/commit) and I've followed
      them.
- [ ] I ran coala over my code locally. (*All commits have to pass
      individually.* It is not sufficient to have "fixup commits" on your PR,
      our bot will still report the issues for the previous commit.) You will
      likely receive a lot of bot comments and build failures if coala does not
      pass on every single commit!

After you submit your pull request, **DO NOT click the 'Update Branch' button.**
When asked for a rebase, consult [coala.io/rebase](https://coala.io/rebase)
instead.

Please consider helping us by reviewing other peoples pull requests as well:

- pick up any PR at <https://coala.io/review>
- review it (check <https://coala.io/reviewing> for more info)
- if you are sure that it needs work, use `corobo mark wip <URL>` to get it out
  of the review queue.

The more you review, the more your score will grow at coala.io and we will
review your PRs faster!

### CODE REVIEW

**Comment by CLiu13** [comment]
Just for reference: alongside the resources provided in https://github.com/coala/coala/issues/5824, I used [this TinyBasic user manual](http://users.telenet.be/kim1-6502/tinybasic/tbum.html) to help create the language definition.

File: coalib/bearlib/languages/definitions/TinyBasic.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+from coalib.bearlib.languages.Language import Language
+
+
+@Language
+class TinyBasic:
+    aliases = 'tb',
+    extensions = '.tb',
+    comment_delimiters = 'REM'
```
Comment (pareksha) [inline coalib/bearlib/languages/definitions/TinyBasic.py]: `comment_delimiters` are a tuple.

File: coalib/bearlib/languages/definitions/TinyBasic.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    string_delimiters = {'"': '"'}
+    encapsulators = {'(': ')'}
+    keywords = [
+        'PRINT', 'IF', 'THEN',
+        'GOTO', 'INPUT', 'LET',
+        'GOSUB', 'RETURN', 'CLEAR',
+        'LIST', 'RUN', 'END',
+        'PR', 'RND', 'USR'
```
Comment (li-boxuan) [inline coalib/bearlib/languages/definitions/TinyBasic.py]: Please add a trailing comma

Comment (CLiu13) [reply coalib/bearlib/languages/definitions/TinyBasic.py]: Added!

File: coalib/bearlib/languages/definitions/TinyBasic.py (hunk 1 above)
Comment (CLiu13) [reply coalib/bearlib/languages/definitions/TinyBasic.py]: Can't believe I glanced over that - fixed!

**Comment by pareksha** [comment]
You need to include your language in https://github.com/coala/coala/blob/master/coalib/bearlib/languages/__init__.py too.

File: coalib/bearlib/languages/definitions/TinyBasic.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
+from coalib.bearlib.languages.Language import Language
+
+
+@Language
+class TinyBasic:
+    aliases = 'tb',
+    extensions = '.tb',
+    comment_delimiters = 'REM',
```
Comment (pareksha) [inline coalib/bearlib/languages/definitions/TinyBasic.py]: I just read that an apostrophe (') can be used in place of the REM command in tiny basic. If it's right, then you need to include it.

Comment (CLiu13) [reply coalib/bearlib/languages/definitions/TinyBasic.py]: I will go check that out, do you have a link that I could look at?

Comment (pareksha) [reply coalib/bearlib/languages/definitions/TinyBasic.py]: I saw in this one http ://p112.sourceforge.net/tbp112.html

It also says that single quotes and double quotes both can be used for character strings. See to this one too.

Comment (CLiu13) [reply coalib/bearlib/languages/definitions/TinyBasic.py]: Nice find, seems legit to me. I will go ahead and read through it to double check for any other discrepancies, and I will push changes soon.

File: coalib/bearlib/languages/definitions/TinyBasic.py [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+
+@Language
+class TinyBasic:
+    aliases = 'tb',
+    extensions = '.tb',
+    comment_delimiters = 'REM', "'"
+    string_delimiters = {'"': '"', "'": "'"}
+    encapsulators = {'(': ')'}
+    keywords = [
```
Comment (Man-Jain) [inline coalib/bearlib/languages/definitions/TinyBasic.py]: I think there are other keywords also like NAP, CLS etc. You can refer here http://tinybc.sourceforge.net

Comment (Man-Jain) [inline coalib/bearlib/languages/definitions/TinyBasic.py]: I think there are other keywords. You can refer here http://tinybasic.de/download/TinyBasicV240.pdf

Comment (CLiu13) [reply coalib/bearlib/languages/definitions/TinyBasic.py]: I think that may be a heavily modified version of TinyBasic designed specifically for programming micro controllers. TinyBasic seems to have been written in the 1970s (see [this guide](http://users.telenet.be/kim1-6502/tinybasic/tbum.html) from 1976), and the guide that you linked to is from 2009. [Wikipedia](https://en.wikipedia.org/wiki/Tiny_BASIC) seems to corroborate the 1970s timeline, and it doesn't refer to any recently updated version of TinyBasic.

In addition to that, the creator of TinyBasic according to Wikipedia is Dennis Allison (from the US), while the creator of the TinyBasic [from the guide that you linked](http://www.tinybasic.de/contact.htm) is Dr.-Ing. (from Germany). The sample code published on [that guide's site](http://www.tinybasic.de/samples.htm) also does not seem to match in style to samples provided in [this original guide](http://users.telenet.be/kim1-6502/tinybasic/tbum.html).

Comment (Man-Jain) [reply coalib/bearlib/languages/definitions/TinyBasic.py]: Ahh that might be the case. What do you say @pareksha . Should we include this ?

Comment (pareksha) [reply coalib/bearlib/languages/definitions/TinyBasic.py]: Tiny Basic is a language that is specifically designed for micro controllers only (that is what I read, correct me if I am wrong).

I see that in the link with Dr. -Ing that the version of language is 2.4.

Read about the new keywords and changes in the language. If there's a major change with the incoming of new versions, include those changes, but with a comment that this change is applicable on and after <version name> version.

Comment (CLiu13) [reply coalib/bearlib/languages/definitions/TinyBasic.py]: Sounds good, I will take care of those changes today.

Comment (CLiu13) [reply coalib/bearlib/languages/definitions/TinyBasic.py]: Changes have been added!

Comment (jayvdb) [reply coalib/bearlib/languages/definitions/TinyBasic.py]: the TinyBasic in that pdf is also umaintained. This is a 2009 snapshot
https://web.archive.org/web/20090410204355/http://www.tinybasic.de/downloads.htm

So I thikn it is not worth considering as part of the TinyBasic history .

**Comment by jayvdb** [comment]
@gitmate-bot rebase

**Comment by jayvdb** [comment]
@gitmate-bot ff

**Pull request merged**